### ISSUE TITLE
Feedback from 9x9 experiment

### ISSUE THREAD

**Issue by Alderi-Tokori** [leela-zero/leela-zero#863]
As of yesterday, minusGo (funny how @amj and I came with such similar names despite being two completely independent projects) got to 9d OGS (https://online-go.com/player/481097/minusGo), which was the "final" milestone I was setting for it. In this post, I'll detail the choices I made concerning training, learning rate, etc. Keep in mind though that this is a 9x9 experiment and the branching factor being much smaller than on 19x19 as well the games being much shorter makes everything work much more easily, and some things that worked here might not work for 19x19 (they'd have to be tested again on 19x19). It's just a little food for thought. The goal of this experiment (aside from me learning things) was to see how far a single machine could train a bot (albeit a 9x9 one) using a zero-ish method.  It is generating games on a server with 2 GTX 1060 cards, while the training and validation matches are done on my personal machine with 2 GTX 1080 cards. The implementation was an original one at first but I quickly switched to Leela Zero because it was getting much faster than mine.

Firstly, here is a google sheet where I maintain the relative strength graph : https://docs.google.com/spreadsheets/d/1X5eaM5fBlgWHPfjM9jAIpu_EV7LFRhVzSHg0F2t9Ud8/edit?usp=sharing
(For those who can't access Google for whatever reason, the up to date strength graph is also on minus' OGS profile page)

During the first 70~80k games, I tested a lot of different things (which resulted in sudden big jumps whenever I found a good thing), but after that (minus was roughly 10k at the time) I've basically kept the same method, which I'll describe here.

The size of the network was (aside from the very first batch of random data) always 6x128, I had tried 9x256 first, then 4x64 and a few other sizes, but in the end I decided to retrain a 6x128 from scratch since it seemed (at the time) to be the best trade-off between speed of game generation and speed of progression.
 Now, with all the optimisations Leela Zero has gone through, a 9x128 is in the realm of possibilities (keep in mind that then I could only generate 1000 games per day with a 6x128, while now I generate over 25'000 games thanks to all those optimisations) once the 6x128 stalls, I suspect that any higher than that won't really make a difference given the level it is already at with a 6x128.

For game generation, I used 1000 playouts per move and resign enabled in 90% of games, the resign threshold was set so that less than 5% of games were incorrect resigns for both colors (sometimes I saw white have 7% wrong resign and black have 1%, so the average was below 5% but I took a smaller threshold so both colors had a below 5% wrong resign rate).

Concerning the size of the training window, I've settled on always only using the games generated by the very latest network after some tests with different window sizes and different "number of latest nets to keep in the window", which seemed to indicate that there was a faster progression with using only the games from the latest only. 

This approach was supposedly the worst in theory in terms of the risk for the network to forget knowledge (and overfitting given the small number of games in the window), but I decided to test it anyway and see if it becomes a problem, since I saw in a reddit AMA of Deepmind that another team had suspicions the MCTS was enough to keep the net from forgetting. Given the net got from 10k to 9d, I think there is little evidence of this happening for now. Another reason I tried this approach is to be free from possibly conflicting data in the training window, and overfit isn't really a problem since at some point there'll be enough games in the window to prevent it.

For the number of steps for each training session, I used the minimum "possible", ergo 2000 steps with a batch size of 512 and a learning rate of 0.005 (equivalent to the first learning rate of 0.01 with a batch size of 2048 that AGZ used). If the 2000-step net didn't get promoted, I tested a 4000-step one. If the 4000-step one failed too, I considered this training session to be a failure, logged the best winrate between the two failed nets in the graph as an unconnected dot, and waited a bit for more games to be generated before trying again. 

At some point around 200k games, I decided to switch to 4000 and 8000 steps because the latest nets were all 4000-step ones. I also decided at the time to switch to -v 1000 instead of -p 1000 in order to boost the speed of the game generation. I surmised at the time that the more than *2 speed-up would counter-balance the "bit worse" data, and so far I don't regret this choice.

As far as learning rate annealing goes, the progression still hasn't stalled, and I'm only at around 140k steps in total right now, so it is still the 0.005 rate of the beginning.

I'll keep the experiment running until it stalls, and then use the games from latest net to bootstrap a 9x128 net that will train using a mix between my current method and the AZ one.

**TL;DR:** Never changed learning rate, only use the fewest number of steps for training, always been 6x128 net (planning on 9x128 one with AZ method), only use the games from latest net instead of a fixed window. Got to 9d in a bit over 600k games and 3 months.

If you have questions, feel free to ask!

**Comment by Marcin1960** [comment]
Is it possible to get the 32bit binary and  a network?

**Comment by Alderi-Tokori** [comment]
@Marcin1960 I'm planning on submitting a PR that will allow people to compile Leelaz for whatever (odd) board size they want, but for now here are the linux binary I'm using + the latest net : https://drive.google.com/file/d/1NRjsri3utQxjTtCN8cQphl2v3G0BZSpj/view?usp=sharing

**Comment by herazul** [comment]
Interesting.

How many time slower would generating selfplay games for 13*13 be ?

**Comment by Marcin1960** [comment]
Of course I would prefer ready 32bit Win binary, as I do not have development platform set.

**Comment by Alderi-Tokori** [comment]
@herazul Hard to say. If we assume the length of the game is related to the board size, then games will be about twice as long, and above that I think the number of calculations for each net eval is related to the board size too, so around twice as much calculations too. So you're looking at a *4 slower figure, though it's a rough estimate, and chances are it won't be that slower (I think between *2 and *4 as slow).

**Comment by Alderi-Tokori** [comment]
@Marcin1960 Ah yeah you want a 32b binary too :/ Can't really help you there then, my Windows binary is 64bit (and it's at home, while I'm at work currently), and even the linux one might be 64bit.

**Comment by Matuiss2** [comment]
There is an  .exe for it?

**Comment by Alderi-Tokori** [comment]
@Matuiss2 There's a 64bit .exe at home, but I could share the freezed code source I'm using for anyone to compile, if that's easier for you.

**Comment by herazul** [comment]
Okay so it might be still very accessible to train a 13*13 network pretty far with a single computer.

**Comment by Matuiss2** [comment]
@Alderi-Tokori I prefer the 64-bit exe, I dont know how to compile(super new in programming here)

**Comment by Alderi-Tokori** [comment]
@Matuiss2 Alright, I'll send another archive with the 64bit exe tonight.

**Comment by Marcin1960** [comment]
What software did you use to compile minusGo? What other components are needed?

**Comment by zediir** [comment]
https://www.visualstudio.com/free-developer-offers/

**Comment by Alderi-Tokori** [comment]
@Marcin1960 At this point minus is just Leelaz with all the hardcoded 19x19 support replaced by a hardcoded 9x9 support. I simply modified what I had to in the source files and compiled it with Visual Studio 2017 (using the msvc project in the repository).

**Comment by Tsun-hung** [comment]
Incredible! I wonder if it can be made progress furtherly? Besides, I want to know if it is stronger than zen5? (9*9 board game)

**Comment by Alderi-Tokori** [comment]
@Tsun-hung I don't have access to Zen so I have no idea, but I don't think it can beat Zen yet. The good thing is it's still showing no sign of slowing down, so it can still get stronger that's for sure.

**Comment by Eddh** [comment]
I'm curious what did ended up with when it comes to FPU ?

**Comment by Alderi-Tokori** [comment]
@Eddh I'm using 0.15 as FPU reduction. I haven't pulled yet since your PR got merged.

**Comment by Eddh** [comment]
I see. Was that an improvement over 0 fpu reduction ?

**Comment by Alderi-Tokori** [comment]
@Eddh I think it was, I went from a 20th Jan code to the early Feb code at once, so I was comparing strength between a binary which still had the TTable code and a binary which did not (and that makes it weaker for the same number of playouts as I've been told), in the end it got around 50% winrate against the old one with 0.15, so I think it'd be stronger than the same binary with FPU = 0, but I haven't tested it.

**Comment by odinAuthor** [comment]
If I understand it correctly only plays with 5.5 komi, any plans to rerun or retrain for 7.0 and test on CGOS 9x9?

**Comment by Alderi-Tokori** [comment]
@odinAuthor I plan to do that yeah, I *think* (that's what I'm interested in testing actually), that I could theoretically suddenly change the komi to 7 and the value net will adjust naturally after a few increments, so I don't lose all the progress it made with shapes on 5.5 komi. If it doesn't work, I'll just go from scratch, but testing the AZ approach.

**Comment by jkiliani** [comment]
Wouldn't a komi of 7 require some major code changes? It would allow for jigo, which means that the current win rate would have to be recoded to mean expected value, similar to what the guys over at https://github.com/glinscott/leela-chess/ are doing...

**Comment by Alderi-Tokori** [comment]
No I don't think it's that big of a change, a jigo would be returned as a value of 0.5, and I'll have to make sure the dump_training function handles that case, but those are pretty minor changes overall.

**Comment by gcp** [comment]
Yup, just return 0.5 for draws.

**Comment by jkiliani** [comment]
From what I see in UCTSearch.cpp, at least the pass heuristics would have to be adjusted for the case that there's jigo on the board. If I read the code correctly, in this case both players will think they win by passing, and end the game in draw even if one could actually win by removing more dead stones.

It's very possible there are more cases in the code that would at least need some looking over if we did account for possibility of draws.

**Comment by Alderi-Tokori** [comment]
@jkiliani In self play (and on CGOS), pass heuristics won't be used, but I get your point, I'll have to be careful if I set a komi that allows for draws.

**Comment by amj** [comment]
@Alderi-Tokori this is awesome!  minigo backed off to a 9x9 run to get better monitoring etc.   I'll write up a full result today and i hope to get the games/models up and hosted for people to look at this week. 

Short version:  We used 32 filters, 9 blocks, no evaluation step and not retraining from scratch each time.   Training window was the most recent 40M positions, which should be ~=500k games.  Games played on a cluster of 800 cpus, 600 playouts/move, resign threshold backed off manually after generation 10 and generation 50, learning rate 0.01, to be cut to 0.001 at 250M positions evaluated (1M steps).

Thanks for sharing your settings and doing this writeup!  Very helpful.  Can i ask what sort of things you were looking at in addition to the selfplay evaluation?  E.g. tensorboard?

**Comment by Alderi-Tokori** [comment]
@amj awesome ^^ I hope you got a good result :) As for what I was looking at, nothing but the results of the validation matches. Of course I was following minus' rank evolution on OGS too, but I didn't use the results there to decide how to proceed.

**Comment by amj** [comment]
@Alderi-Tokori Well, it's beating me and the other dans in the office so... so far so good ;)  I think we'd have to do some more legwork to put it up on OGS, and since we trained with 7.5 komi, CGOS won't work.

**Comment by Alderi-Tokori** [comment]
@Matuiss2 Here is the link with the win64 binary, (hopefully) all the needed dlls, and (again) the current best network. First time I distribute that so I really have no idea if it will run on your computer though: https://drive.google.com/file/d/14h4OCfpeJwrZV-nPlfbYkRcstLVYWGcf/view?usp=sharing

**Comment by Matuiss2** [comment]
@Alderi-Tokori Thanks it worked

**Comment by Alderi-Tokori** [comment]
On OGS it runs with -t 2 -v 16000, with two gpus it means 10s per move maximum.

**Comment by Matuiss2** [comment]
It seems like 5.5 komi is very black color favored, i ran 100 games vs Crazystone 7d and only 2 games were won by white.

**Comment by odinAuthor** [comment]
I play 9x9 correspondence tournaments with Valkyria on OGS and LittleGolem (using up to 24 hours cpu MCTS per move). To my experience black with less than 7.0 komi should be a free win. But Valkyria still has enough bugs and biases, so it can be beaten despite am opening book built up over many years and long thinking times. Therefor I think in the long run 7.0 komi is more interesting. On CGOS jigo is still the least likely outcome between the top bots. Perhaps the reason is that strong 9x9 play often leads to very long games with semeai/ko/LD-fights where jigo is not possible.

**Comment by Alderi-Tokori** [comment]
@Matuiss2 From self play, minus wins around 55% of the games as black with 5.5 komi. As for the games against Crazystone, mind sharing the final results and what setting you used? I'm curious if minus is already on the level of the commercial AIs or not.

**Comment by Matuiss2** [comment]
@Alderi-Tokori Ill start taking notes, but from what Im seeing, black always wins no matter who plays it the few ones that white won was crazystone playing as white, so Crazystone should be very slightly better. but ill make a more detailed report later(leela is playing with 12k playouts)

**Comment by Alderi-Tokori** [comment]
@Matuiss2 Nice, I didn't think it had a shot at CS yet, hopefully in a few increments it will the one to win as white :p

**Comment by KyleBiedermann** [comment]
Minusgo does tend to play the same thing over and over so some experiments may be a little difficult on reliability as minus's openings tend to vary depending on the new nets for better or worse. I usually play many different moves with white and black to get a clearer understanding if it stronger than the program in some aspects while at the same time expanding the database and removing flawed data to get a better understand if 7 komi lead to perfect play that and some of the lines it plays can sometimes be not so good especially with white.

**Comment by Alderi-Tokori** [comment]
@KyleBiedermann I'm guessing you're Deep Scholar's maintainer? Keep up the good work! :)

**Comment by Marcin1960** [comment]
I checked Visual Studio package. Unfortunately it is not for me, it is more than one gigabyte in size and installation of such package just set one compilation flag differently does not make sense.

So I wish you success.

**Comment by Eddh** [comment]
Does someone know who is running the cgos bot mentioned here ? http://computer-go.org/pipermail/computer-go/2018-February/010704.html

**Comment by Alderi-Tokori** [comment]
@Eddh No idea, not me.

**Comment by KyleBiedermann** [comment]
@Alderi-Tokori Thanks Alderi the work can be very tricky as currently working on most my openings with black as some seem flawed or lacking knowledge as I have changed it to 5-4 due to more reasonable and solid results on 5.5 results. So the labor of hand punching in new lines will take some time to come to fruition. >>

**Comment by odinAuthor** [comment]
The author of the unknown bot on 9x9 CGOS has now created a new one "CLeelaz9Test1" which is much stronger than the original. Maybe it is using a supervised network trained on CGOS games. This is also what I am doing with "ZOdin_12_C...". It started as validation of my Zero code but now I experimenting with smaller and faster networks and it looks good. For the pure Zero version I am only doing 7x7 to find working hyperparameters for small boards.

**Comment by FFLaguna** [comment]
Do you by any chance have an updated parse.py and tfprocess.py in order to train this or another network further on TF?

**Comment by Matuiss2** [comment]
@Alderi-Tokori 
Finished mine crazystone 5 secs-move vs minusgo 16000 playouts 110 games match
crazystone won 44 games 
leelaz9x9 won 66 games
black won 79 games
white won 31 games
Its a small sample but its clear that this bot is stronger than crazystone at 16000 playouts, and black is more than 55% favored at 5.5 komi.

**Comment by LetterRip** [comment]
An experiment I'd like to see, after you've made as much progress as you'd care to.  Is to then do a series of training games that all start with popular opening sequences (perhaps frequency of the sequence by the popularity in pro and amateur games), and then train the bot on those.

**Comment by Alderi-Tokori** [comment]
@FFLaguna Here is the link to the parse.py and tfprocess I'm using, it arranges games in the given directory in chunks of 128 in the "chunks" directory if it doesn't exist yet (if it already exists, it skips this process), then launch the training on those. It's not the latest files so there's no true validation split in it: https://drive.google.com/file/d/1UPSh1zWXPUoqz112aG-nGORT16AJRn1O/view?usp=sharing

@Matuiss2 Interesting, but you gave more time to Leelaz than to CS in your state, could you rerun the test with -t 2 -p 8000 (should be around 5s per move) ?

@LetterRip I don't really see any reason to run this experiment, to be honest, much more so now that you have the binaries, weight file, and training scripts to do it yourself (and since I've done everything on one machine, you can too).

**Comment by KyleBiedermann** [comment]
@Matuiss2 Very interesting sample size indeed, the other option to Alderi recommendation is to increase the time & playouts of Crazystone to graph the performance and the diminishing returns of more playouts against the leela zero 9x9.

**Comment by d7urban** [comment]
@Matuiss2 How are you doing the Crazy Stone matches? Is there an engine-only version available somewhere that I could get my hands on?

**Comment by Matuiss2** [comment]
@d7urban im doing then by hand about a game every 5-6 minutes

**Comment by Alderi-Tokori** [comment]
@Matuiss2 Thanks a lot for putting time into this :)

**Comment by Marcin1960** [comment]
After all, I installed Visual Studio. Is the source available?

**Comment by Alderi-Tokori** [comment]
Not in a proper way (ergo not through a PR), but here's my current source, which is a bit behind the current /next : https://drive.google.com/file/d/1VMBuMMGKywbyCOz8gDpjZa_NGgjmSnsl/view?usp=sharing

**Comment by Marcin1960** [comment]
Thank you very much.

**Comment by ayssia** [comment]
@Alderi-Tokori I wonder if there are any new 9x9 models and how they performs?

**Comment by Alderi-Tokori** [comment]
@ayssia I did a two weeks pause in progress in order to test the rate of progression of FPU reduction against no reduction, but now I've started generating new nets. For the record, it seems the original promotion to 9d was a fluke (a 7d that wasn't actually 7d apparently got a few loss in a row against minus and that pushed it over the 9d threshold). Actually it's probable minus was a strong 7d / weak 8d at the time.

There have been 14 promotions since then and it doesn't really seem to have gotten stronger despite the nets promoting easily each time. That's what originally prompted me to do my no reduction VS reduction test to see if it was the cause, but it doesn't seem to be it. My current guess is either:
- this is due to my choice to use only the latest games
- this is due to the net saturating in a way, making a rock-paper-scissor dynamic appear in promotions, while the actual strength stays the same.

I'm almost at 1M games generated, and I'll keep running minus until at least 1.5M games. If at that time it still hasn't done any progress, I'll start anew with a fixed window.

**Comment by Marcin1960** [comment]
"miniGo was 9x32 while minusGo is 6x128"

I would suggest 10x64. That way you are assured, that it will be stronger than than miniGo. :)

**Comment by Alderi-Tokori** [comment]
@Marcin1960 My current experiment is 12x32, it appears much more isn't really needed if miniGo got to pro level with a 9x32 net.

**Comment by Alderi-Tokori** [comment]
Keep in mind that now the source code and binaries I provided are obsolete, and you can compile current /next Leela Zero for 9x9 by simply changing a constant (BOARD_SIZE) in config.h

**Comment by alreadydone** [comment]
I seem to remember even height or even width boards are not yet supported. Can anyone remind me what the difficulty is? I am wondering whether the policy network will produce very good move ordering for principal variation search and facilitate the proof of the correct komis on small rectangular or toroidal boards. (Toroidal boards should use cyclic padding instead of zero padding, and may not need the fully connected layer because of strict translational symmetry.)

**Comment by Alderi-Tokori** [comment]
@alreadydone The problem is several parts of the code (IIRC Winograd tiles and training data dumping) assume an odd board size.

**Comment by Ttl** [comment]
Winograd convolution should work with even board sizes. Leela-chess is actually using the same code on 8x8 chess board.

**Comment by Marcin1960** [comment]
Is there a newer, stronger 9x9 net than February 12?

**Comment by Alderi-Tokori** [comment]
@Marcin1960 There is one but it's not that much stronger (which is the reason I did https://github.com/gcp/leela-zero/issues/863#issuecomment-372243689 ), you're good with the one provided here.

**Comment by wensdong** [comment]
where is the source code?

**Comment by Marcin1960** [comment]
@wensdong 

It is the same code. Just change 19 to 9 in few places. I tried it and it works.

**Comment by Alderi-Tokori** [comment]
@Marcin1960 No. It is the same code, just change 19 to 9 in a *single* place.

**Comment by wensdong** [comment]
@Alderi-Tokori
Is it this line   "#define BOARD_SIZE 19"  in config.h ?  I did not realise the code is now configurable, tried to do it myself. Thanks!

**Comment by Alderi-Tokori** [comment]
@wensdong Yup, that's what you need to change.

**Comment by wensdong** [comment]
@Alderi-Tokori thanks you so much for making the effort.  Now I can explore the algorithm further!

**Comment by Alderi-Tokori** [comment]
@wensdong Have fun :)

**Comment by Marcin1960** [comment]
Well, I changed also 19 to 9 in SGFTree.cp. Possibly it does not matter.

**Comment by Alderi-Tokori** [comment]
@Marcin1960 The 19s in SGFTree.cpp  are in the function to create a game tree from a sgf file, when no size is specified in the file, the default size according to the SGF specs is assumed and that's 19.

**Comment by wensdong** [comment]
@Alderi-Tokori  

Sorry to bother you again. Do you have a initial weight file? I used one, but run into trouble.

**Comment by Alderi-Tokori** [comment]
@wensdong Initial weight as in the random ones? You can generate them with the training scripts if I'm not mistaken.

**Comment by wensdong** [comment]
@Alderi-Tokori  which training scripts?

**Comment by Alderi-Tokori** [comment]
@wensdong training/tf/, but you'd have to tweak the code a bit to construct and save directly a new net.

**Comment by urza** [comment]
Hello, sorry to hijack this old thread, dont know where or how else to contant @Alderi-Tokori 
What is the best way to get your minusGo bot today for selft study and play? I know I can play on online-go.com, but I would like to for example give it a 9x9 board situation and let it evaluate best moves etc.
Is it possible to use it as a dataset for current leela-zero release? Is it downloadable somewhere? The links in this thread no longer work..
Thanks a lot

**Comment by alreadydone** [comment]
https://github.com/ihavnoid/leelaz-ninenine/issues/6

**Comment by Alderi-Tokori** [comment]
Hey @urza, 

The link Alreadydone provided contains stronger nets than minus iirc, so if you want the best of the best, you're better off using these. Regardless, I'll provide an updated download link for minus' weights tomorrow.

**Comment by urza** [comment]
@Alderi-Tokori looking forward. I find current minusgo gamestyle particularly aesthetic and pleasing in some ways
How many visits is it configured to on online-go? Or is it dynamic depending on time left?

@Alreadydone thanks, looks like very strong network

**Comment by iopq** [comment]
@Alderi-Tokori 158s.gz is stronger than the ones in that thread

**Comment by Alderi-Tokori** [comment]
Ah? I don't monitor the github as closely as back then so I must have missed it. Anyway @urza, here's an updated link: https://drive.google.com/open?id=1mCoHG_ztHtMMMx7RYO5mfyiZk9wjmeAl

The command arguments I use on minusGo are "leelaz -g -r 30 -w leelaz-model-best --gpu 0 --gpu 1 -t 2 -v 200000 --fpu_reduction 0.15", you need 2 gpus for that, though. Otherwise you can use "leelaz -g -r 30 -w leelaz-model-best --gpu 0 -t 2 -v 200000 --fpu_reduction 0.15", it will simply be slower.

**Comment by iopq** [comment]
What number network is that?

**Comment by Alderi-Tokori** [comment]
That's number 63.

**Comment by urza** [comment]
@Alderi-Tokori Amazing, thank you very much

**Comment by hzyhhzy** [comment]
Update:
komi 7.5 169th weight(169s.txt.gz)30b320f about 4.5M games(50visit then 35po then 50po)
https://drive.google.com/open?id=1mxV-ERlmnARfM--9RFHLI0eTpyh3SwPD
komi 6.5 127th weight(65_127s.txt.gz)20b256f about 2M games(25po)
https://drive.google.com/open?id=1YTsV4nwmewYooypyPaYP7I8vei46H0x6

**Comment by hzyhhzy** [comment]
A fast rough test:
169s 1po vs leelaz-model-best 200po  52:16
![屏幕截图(21)](https://user-images.githubusercontent.com/41548394/58689121-bd9e4c80-83b8-11e9-8fba-3bbc22b01d9d.png)

**Comment by Alderi-Tokori** [comment]
Nicely done @hzyhhzy :) I need to get back into this. I want to try to do the SAI approach but I haven't got a server anymore on which to run the training and game generation :p
